java-frontend/src/main/java/org/sonar/java/bytecode/cfg/BytecodeCFGBuilder.java

```diff
@@ -0,0 +1,234 @@
+    } catch (IOException e) {
+      throw Throwables.propagate(e);
+    } finally {
+      Closeables.closeQuietly(is);
+    }
+  }
+
+  static class CFGBytecode {
+    List<Block> blocks;
+    Symbol.MethodSymbol methodSymbol;
+
+    CFGBytecode(Symbol.MethodSymbol methodSymbol) {
+      this.methodSymbol = methodSymbol;
+      blocks = new ArrayList<>();
+      // create exit block
+      blocks.add(new Block(this));
+    }
+  }
+  static class Block {
+    int id;
+    CFGBytecode cfg;
+    List<Instruction> instructions;
+    List<Block> successors;
+
+    Block(CFGBytecode cfg) {
+      this.cfg = cfg;
+      this.id = cfg.blocks.size();
+      instructions = new ArrayList<>();
+      successors = new ArrayList<>();
+    }
+
+    void addInsn(int opcode) {
+      instructions.add(new Instruction(opcode));
+    }
+
+    Block createSuccessor() {
+      Block newBlock = new Block(cfg);
+      successors.add(newBlock);
+      cfg.blocks.add(newBlock);
+      return newBlock;
+    }
+
+    String printBlock() {
+      StringBuilder sb = new StringBuilder();
+      sb.append("B").append(id);
+      if (id == 0) {
+        sb.append("(Exit)\n");
+        return sb.toString();
+      }
+      sb.append("\n");
+      int index = 0;
+      for (Instruction instruction : instructions) {
+        sb.append(index).append(": ").append(Printer.OPCODES[instruction.opcode]).append("\n");
+        index++;
+      }
+      sb.append("Jumps to: ");
+      successors.forEach(s -> sb.append("B").append(s.id).append(" "));
+      sb.append("\n");
+      return sb.toString();
+    }
+  }
+
+  private static class BytecodeCFGMethodVisitor extends MethodVisitor {
+    Map<Label, Block> blockByLabel = new HashMap<>();
+    private Block currentBlock;
+    private CFGBytecode cfg;
+
+    BytecodeCFGMethodVisitor(Symbol.MethodSymbol methodSymbol) {
+      super(Opcodes.ASM5);
+      cfg = new CFGBytecode(methodSymbol);
+      currentBlock = new Block(cfg);
+    }
+
+    @Override
+    public void visitInsn(int opcode) {
+      currentBlock.addInsn(opcode);
+    }
+
+    @Override
+    public void visitIntInsn(int opcode, int operand) {
+      currentBlock.addInsn(opcode);
+    }
+
+    @Override
+    public void visitVarInsn(int opcode, int var) {
+      currentBlock.addInsn(opcode);
+    }
+
+    @Override
+    public void visitFieldInsn(int opcode, String owner, String name, String desc) {
+      currentBlock.addInsn(opcode);
+    }
+
+    @Override
+    public void visitMethodInsn(int opcode, String owner, String name, String desc, boolean itf) {
+      currentBlock.addInsn(opcode);
+    }
+
+    @Override
+    public void visitInvokeDynamicInsn(String name, String desc, Handle bsm, Object... bsmArgs) {
+      currentBlock.addInsn(Opcodes.INVOKEDYNAMIC);
+    }
+
+    @Override
+    public void visitLdcInsn(Object cst) {
+      // FIXME (npe) LDC is generic information about constant but the real value of constant could be loaded into instructions
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Take the required action to fix the issue indicated by this comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1134)